File: Mac/AWDLLinkInfo.m
#import "AWDLLinkInfo.h"
#import <objc/message.h>
#import <objc/runtime.h>
#import <dlfcn.h>

static id _Nullable ODObjectGetterWithError(id object, const char *selectorName) {
    SEL selector = sel_registerName(selectorName);
    if (!object || ![object respondsToSelector:selector]) return nil;
    NSError *__autoreleasing error = nil;
    id (*send)(id, SEL, NSError *__autoreleasing *) = (void *)objc_msgSend;
    id value = send(object, selector, &error);
    return error == nil ? value : nil;
}

static NSNumber * _Nullable ODNumber(id value) {
    if ([value isKindOfClass:NSNumber.class]) return value;
    if ([value isKindOfClass:NSString.class]) {
        NSScanner *scanner = [NSScanner scannerWithString:value];
        double number = 0;
        if ([scanner scanDouble:&number]) return @(number);
    }
    return nil;
}

NSDictionary<NSString *, id> * _Nullable ODCopyAWDLLinkInfo(void) {
    static void *coreWiFi = NULL;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        coreWiFi = dlopen("/System/Library/PrivateFrameworks/CoreWiFi.framework/CoreWiFi",
                          RTLD_LAZY | RTLD_LOCAL);
    });
    if (!coreWiFi) return nil;

    @try {
        Class cls = NSClassFromString(@"CWFApple80211");
        if (!cls) return nil;
        id (*allocSend)(id, SEL) = (void *)objc_msgSend;
        id allocated = allocSend((id)cls, sel_registerName("alloc"));
        id (*initSend)(id, SEL, id) = (void *)objc_msgSend;
        id radio = initSend(allocated, sel_registerName("initWithInterfaceName:"), @"awdl0");
        if (!radio) return nil;

        NSMutableDictionary<NSString *, id> *result = [NSMutableDictionary dictionary];

        id channelObject = ODObjectGetterWithError(radio, "channel:");
        if (channelObject) {
            SEL channelSelector = sel_registerName("channel");
            SEL bandSelector = sel_registerName("band");
            SEL widthSelector = sel_registerName("width");
            if ([channelObject respondsToSelector:channelSelector]) {
                unsigned long long (*send)(id, SEL) = (void *)objc_msgSend;
                unsigned long long channel = send(channelObject, channelSelector);
                if (channel > 0 && channel <= NSIntegerMax) result[@"channel"] = @((NSInteger)channel);
            }
            if ([channelObject respondsToSelector:bandSelector]) {
                unsigned (*send)(id, SEL) = (void *)objc_msgSend;
                result[@"band"] = @(send(channelObject, bandSelector));
            }
            if ([channelObject respondsToSelector:widthSelector]) {
                int (*send)(id, SEL) = (void *)objc_msgSend;
                int width = send(channelObject, widthSelector);
                if (width > 0) result[@"widthMHz"] = @(width);
            }
            result[@"channelDescription"] = channelObject.description ?: @"";
        }

        struct { const char *selector; NSString *key; } numberGetters[] = {
            { "txRate:", @"txRateMbps" },
            { "rxRate:", @"rxRateMbps" },
            { "maxLinkSpeed:", @"maxLinkMbps" },
            { "MCSIndex:", @"mcs" },
            { "RSSI:", @"rssi" },
        };
        for (size_t i = 0; i < sizeof(numberGetters) / sizeof(numberGetters[0]); i++) {
            NSNumber *number = ODNumber(ODObjectGetterWithError(radio, numberGetters[i].selector));
            if (number) result[numberGetters[i].key] = number;
        }

        id phy = ODObjectGetterWithError(radio, "activePHYMode:");
        if (phy) {
            NSNumber *number = ODNumber(phy);
            if (number) result[@"phyMode"] = number;
            result[@"phyDescription"] = phy.description ?: @"";
        }

        id master = ODObjectGetterWithError(radio, "AWDLMasterChannel:");
        if (NSNumber *number = ODNumber(master)) result[@"masterChannel"] = number;
        id secondary = ODObjectGetterWithError(radio, "AWDLSecondaryMasterChannel:");
        if (NSNumber *number = ODNumber(secondary)) result[@"secondaryMasterChannel"] = number;

        id sequence = ODObjectGetterWithError(radio, "AWDLSyncChannelSequence:");
        if (sequence) result[@"channelSequence"] = sequence.description ?: @"";

        return result.count > 0 ? result : nil;
    } @catch (__unused NSException *exception) {
        // Private diagnostics must never affect streaming. API drift simply
        // means the AWDL detail row disappears on that OS build.
        return nil;
    }
}

File: Mac/AWDLLinkInfo.swift
import Foundation

struct AWDLLinkInfo: Equatable, Sendable {
    let channel: Int?
    let band: Int?
    let bandwidthMHz: Int?
    let frequencyMHz: Int?
    let txRateMbps: Double?
    let rxRateMbps: Double?
    let maxLinkMbps: Double?
    let mcs: Int?
    let rssi: Int?
    let phyMode: Int?
    let masterChannel: Int?
    let secondaryMasterChannel: Int?
    let channelSequence: String?

    var bandLabel: String? {
        switch band {
        case 1: return "2.4 GHz"
        case 2: return "5 GHz"
        case 3: return "6 GHz"
        default: return nil
        }
    }

    static func frequencyMHz(channel: Int, band: Int) -> Int? {
        guard channel > 0 else { return nil }
        switch band {
        case 1:
            if channel == 14 { return 2484 }
            guard (1...13).contains(channel) else { return nil }
            return 2407 + 5 * channel
        case 2:
            return 5000 + 5 * channel
        case 3:
            // 6 GHz normally follows 5950 + 5*n; channel 2 at 5935 MHz is
            // the standardized special case.
            if channel == 2 { return 5935 }
            return 5950 + 5 * channel
        default:
            return nil
        }
    }

    static func parse(_ raw: [AnyHashable: Any]) -> AWDLLinkInfo? {
        func int(_ key: String) -> Int? {
            (raw[key] as? NSNumber)?.intValue ?? raw[key] as? Int
        }
        func double(_ key: String) -> Double? {
            (raw[key] as? NSNumber)?.doubleValue ?? raw[key] as? Double
        }
        let channel = int("channel")
        let band = int("band")
        let info = AWDLLinkInfo(
            channel: channel,
            band: band,
            bandwidthMHz: int("widthMHz"),
            frequencyMHz: channel.flatMap { c in band.flatMap { Self.frequencyMHz(channel: c, band: $0) } },
            txRateMbps: double("txRateMbps"),
            rxRateMbps: double("rxRateMbps"),
            maxLinkMbps: double("maxLinkMbps"),
            mcs: int("mcs"),
            rssi: int("rssi"),
            phyMode: int("phyMode"),
            masterChannel: int("masterChannel"),
            secondaryMasterChannel: int("secondaryMasterChannel"),
            channelSequence: raw["channelSequence"] as? String
        )
        let hasUsefulValue = info.channel != nil || info.txRateMbps != nil || info.rxRateMbps != nil
            || info.maxLinkMbps != nil || info.rssi != nil
        return hasUsefulValue ? info : nil
    }

    static func detect() async -> AWDLLinkInfo? {
        await Task.detached(priority: .utility) {
            guard let raw = ODCopyAWDLLinkInfo() as? [AnyHashable: Any] else { return nil }
            return parse(raw)
        }.value
    }
}
